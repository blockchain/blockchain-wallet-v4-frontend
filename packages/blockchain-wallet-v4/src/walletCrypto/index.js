import * as crypto from 'crypto'
import { pbkdf2 } from 'pbkdf2'
import assert from 'assert'
import Task from 'data.task'
import BIP39 from 'bip39'
import * as U from './utils'
import Either from 'data.either'
import createRng from './rng'
import { curry, compose, is, isNil, sequence, has, propSatisfies } from 'ramda'

const SUPPORTED_ENCRYPTION_VERSION = 3

// eitherToTask :: Either a b -> Task a b
const eitherToTask = e => e.fold(Task.rejected, Task.of)

// TaskTry :: (a -> b) -> (a -> Task Error b)
const TaskTry = f =>
  compose(
    eitherToTask,
    Either.try(f)
  )

// satisfy :: [Bool, String] => Task Error Bool
const satisfy = ([bool, message]) =>
  bool ? Task.of(true) : Task.rejected(message)

// satisfyAll :: [[Bool, String]] => Task Error Bool
const satisfyAll = conditions => sequence(Task.of, conditions.map(satisfy))

// safeParse :: String -> String -> Task error JSON
export const safeParse = (payload, errorMessage) => {
  try {
    return Task.of(JSON.parse(payload))
  } catch (e) {
    return Task.rejected(errorMessage || e)
  }
}

// toPayloadAndIV :: Buffer -> Object
const toPayloadAndIV = data => {
  let dataHex = Buffer.from(data, 'base64')
  let iv = dataHex.slice(0, U.SALT_BYTES)
  let payload = dataHex.slice(U.SALT_BYTES)
  return { payload, iv }
}

// pbkdf2Task :: String -> String -> Integer -> Integer -> Task Error Buffer
const pbkdf2Task = (password, salt, iterations, keyLenBits) => {
  try {
    const saltBuffer = Buffer.from(salt, 'hex')
    const keyLenBytes = (keyLenBits || 256) / 8
    return new Task((reject, resolve) => {
      let handle = (error, key) => {
        if (error) reject(error)
        else resolve(key)
      }
      pbkdf2(password, saltBuffer, iterations, keyLenBytes, 'sha1', handle)
    })
  } catch (e) {
    return Task.rejected(e)
  }
}

// payload: (Buffer)
// iv: initialization vector (Buffer)
// key: AES key (256 bit Buffer)
// options: (optional)
// returns: decrypted payload (e.g. a JSON string)
const decryptBufferWithKey = (payload, iv, key, options) => {
  options = options || {}
  options.padding = options.padding || U.Iso10126
  let decryptedBytes = U.AES.decrypt(payload, key, iv, options)
  return decryptedBytes.toString('utf8')
}

// hashNTimes :: Integer -> String -> String
export const hashNTimes = curry((iterations, data) => {
  assert(iterations > 0, '`iterations` must be a number greater than 0')
  let result = data
  for (let i = 1; i <= iterations; i++) {
    result = sha256(result)
  }
  return result
})

// isStringHashInFraction :: String -> Float -> Bool
export const isStringHashInFraction = (str, fraction) => {
  if (!str) return false
  if (fraction < 0) return false
  return sha256(str)[0] / 256 >= 1 - fraction
}

// /////////////////////////////////////////////////////////////////////////////

// data: e.g. JSON.stringify({...})
// key: AES key (256 bit Buffer)
// iv: optional initialization vector
// returns: concatenated and Base64 encoded iv + payload
export const encryptDataWithKey = curry((data, key, iv) => {
  let IV = iv || crypto.randomBytes(U.SALT_BYTES)
  let dataBytes = Buffer.from(data, 'utf8')
  let options = { mode: U.AES.CBC, padding: U.Iso10126 }
  let encryptedBytes = U.AES.encrypt(dataBytes, key, IV, options)
  let payload = Buffer.concat([IV, encryptedBytes])
  return payload.toString('base64')
})

// data: base64 string
// key: AES key (256 bit Buffer)
// returns; decrypted payload (e.g. a JSON string)
export const decryptDataWithKey = curry((data, key) => {
  let dataHex = Buffer.from(data, 'base64')
  let iv = dataHex.slice(0, U.SALT_BYTES)
  let payload = dataHex.slice(U.SALT_BYTES)
  return decryptBufferWithKey(payload, iv, key)
})

// sha256 :: Buffer -> Buffer
export const sha256 = data =>
  crypto
    .createHash('sha256')
    .update(data)
    .digest()

// generateMnemonic :: Api -> Promise String
export const generateMnemonic = api => {
  return createRng(16, api).then(rng => BIP39.generateMnemonic(null, rng))
}

// stretchPassword :: password -> salt -> iterations -> keylen -> Task Error Buffer
export const stretchPassword = (password, salt, iterations, keyLenBits) => {
  return satisfyAll([
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required'],
    [salt, 'salt_required'],
    [
      isNil(keyLenBits) || keyLenBits % 8 === 0,
      'key_len_multiple_of_8_required'
    ]
  ]).chain(() => pbkdf2Task(password, salt, iterations, keyLenBits))
}

// decryptDataWithPassword :: data -> password -> iterations -> options -> Task Error Buffer
export const decryptDataWithPassword = (
  data,
  password,
  iterations,
  options
) => {
  if (!data) return Task.of(data)
  return satisfyAll([
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required']
  ]).chain(() =>
    TaskTry(toPayloadAndIV)(data).chain(({ iv, payload }) =>
      stretchPassword(password, iv, iterations, U.KEY_BIT_LEN).chain(key =>
        TaskTry(decryptBufferWithKey)(payload, iv, key, options)
      )
    )
  )
}

// encryptDataWithPassword :: String -> String -> Integer -> Task, Error String
export const encryptDataWithPassword = (data, password, iterations) => {
  if (!data) return Task.of(data)
  let salt = crypto.randomBytes(U.SALT_BYTES)
  return satisfyAll([
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required']
  ])
    .chain(() => stretchPassword(password, salt, iterations, U.KEY_BIT_LEN))
    .chain(key => TaskTry(encryptDataWithKey)(data, key, salt))
}

// encryptSecPass :: String -> Integer -> String -> String -> Task, Error String
export const encryptSecPass = curry(
  (sharedKey, pbkdf2Iterations, password, message) =>
    encryptDataWithPassword(message, sharedKey + password, pbkdf2Iterations)
)

const checkFailure = curry((pass, fail, str) =>
  str === '' ? fail(new Error('DECRYPT_FAILURE')) : pass(str)
)

// decryptSecPass :: String -> Integer -> String -> String -> Task, Error String
export const decryptSecPass = curry(
  (sharedKey, pbkdf2Iterations, password, message) =>
    decryptDataWithPassword(
      message,
      sharedKey + password,
      pbkdf2Iterations
    ).chain(checkFailure(Task.of, Task.rejected))
)

// TODO :: review users of that funciton (moved to task)
// encryptWallet :: String -> String -> Integer -> String -> Task Error String
export const encryptWallet = curry((data, password, iterations, version) =>
  satisfyAll([
    [data, 'data_required'],
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required'],
    [version, 'version_required']
  ])
    .chain(() => encryptDataWithPassword(data, password, iterations))
    .map(payload =>
      JSON.stringify({ pbkdf2_iterations: iterations, version, payload })
    )
)

// decryptWalletV1 :: String -> String -> Task Error Object
export const decryptWalletV1 = (password, data) => {
  let decrypt = (i, o) =>
    decryptDataWithPassword(data, password, i, o).chain(
      safeParse,
      'v1: wrong_wallet_password'
    )
  // v1: CBC, ISO10126, 10 iterations
  return (
    decrypt(10)
      // v1: OFB, nopad, 1 iteration
      .orElse(decrypt.bind(null, 1, { mode: U.AES.OFB, padding: U.NoPadding }))
      // v1: OFB, ISO7816, 1 iteration
      // ISO/IEC 9797-1 Padding method 2 is the same as ISO/IEC 7816-4:2005
      .orElse(decrypt.bind(null, 1, { mode: U.AES.OFB, padding: U.Iso97971 }))
      // v1: CBC, ISO10126, 1 iteration
      .orElse(decrypt.bind(null, 1, { mode: U.AES.CBC, padding: U.Iso10126 }))
  )
}

const validateWrapper = (
  wrapper // wrapper
) =>
  satisfyAll([
    [has('payload', wrapper), 'v2v3 wrapper: payload_required'],
    [has('version', wrapper), 'v2v3 wrapper: version_required'],
    [
      has('pbkdf2_iterations', wrapper),
      'v2v3 wrapper: pbkdf2_iterations_required'
    ],
    [
      propSatisfies(v => v <= SUPPORTED_ENCRYPTION_VERSION, 'version', wrapper),
      'Wallet version ' + wrapper.version + ' not supported.'
    ]
  ]).map(() => wrapper)

// decryptWalletV2V3 :: String -> String -> Task Error Object
export const decryptWalletV2V3 = (password, data) => {
  return (
    safeParse(data, 'v2v3: wrong_wrapper')
      .chain(validateWrapper)
      // v2/v3: CBC, ISO10126, iterations in wrapper
      .chain(w =>
        decryptDataWithPassword(w.payload, password, w.pbkdf2_iterations)
      )
      .chain(p => safeParse(p, 'v2v3: wrong_wallet_password'))
  )
}

export const decryptWallet = curry((password, data) =>
  decryptWalletV1(password, data).orElse(() =>
    decryptWalletV2V3(password, data)
  )
)
