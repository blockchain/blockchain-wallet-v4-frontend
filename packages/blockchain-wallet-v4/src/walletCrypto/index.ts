import assert from 'assert'
import BigInteger from 'bigi'
import BIP39 from 'bip39'
import * as Bitcoin from 'bitcoinjs-lib'
import * as crypto from 'crypto'
import Either from 'data.either'
import Task from 'data.task'
import * as curve from 'ecurve'
import hkdf from 'futoin-hkdf'
import { pbkdf2 } from 'pbkdf2'
// @ts-ignore
import { compose, curry, has, is, isNil, propSatisfies, sequence } from 'ramda'

import createRng from './rng'
import * as U from './utils'

export const SUPPORTED_ENCRYPTION_VERSION = 4

// eitherToTask :: Either a b -> Task a b
const eitherToTask = (e) => e.fold(Task.rejected, Task.of)

// TaskTry :: (a -> b) -> (a -> Task Error b)
const TaskTry = (f) => compose(eitherToTask, Either.try(f))

// satisfy :: [Bool, String] => Task Error Bool
const satisfy = ([bool, message]) => (bool ? Task.of(true) : Task.rejected(message))

// satisfyAll :: [[Bool, String]] => Task Error Bool
const satisfyAll = (conditions) => sequence(Task.of, conditions.map(satisfy))

// safeParse :: String -> String -> Task error JSON
export const safeParse = (payload, errorMessage) => {
  try {
    return Task.of(JSON.parse(payload))
  } catch (e) {
    return Task.rejected(errorMessage || e)
  }
}

// toPayloadIV :: Buffer -> Object
const toPayloadIV = (data) => {
  const dataHex = Buffer.from(data, 'base64')

  const iv = dataHex.slice(0, U.SALT_BYTES)
  const payload = dataHex.slice(U.SALT_BYTES)
  return { iv, payload }
}

// pbkdf2Task :: String -> String -> Integer -> Integer -> Task Error Buffer
const pbkdf2Task = (password, salt, iterations, keyLenBits) => {
  try {
    const saltBuffer = Buffer.from(salt, 'hex')
    const keyLenBytes = (keyLenBits || 256) / 8
    return new Task((reject, resolve) => {
      const handle = (error, key) => {
        if (error) reject(error)
        else resolve(key)
      }
      pbkdf2(password, saltBuffer, iterations, keyLenBytes, 'sha1', handle)
    })
  } catch (e) {
    return Task.rejected(e)
  }
}

// payload: (Buffer)
// iv: initialization vector (Buffer)
// key: AES key (256 bit Buffer)
// tag: AES.CBC authTag
// options: (optional)
// returns: decrypted payload (e.g. a JSON string)
const decryptBufferWithKey = (payload, iv, key, options) => {
  options = options || {}
  options.padding = options.padding || U.Iso10126
  const decryptedBytes = U.AES.decrypt(payload, key, iv, options)
  return decryptedBytes.toString('utf8')
}

// hashNTimes :: Integer -> String -> String
export const hashNTimes = curry((iterations, data) => {
  assert(iterations > 0, '`iterations` must be a number greater than 0')
  let result = data
  for (let i = 1; i <= iterations; i++) {
    result = sha256(result)
  }
  return result
})

// isStringHashInFraction :: String -> Float -> Bool
export const isStringHashInFraction = (str, fraction) => {
  if (!str) return false
  if (fraction < 0) return false
  return sha256(str)[0] / 256 >= 1 - fraction
}

// /////////////////////////////////////////////////////////////////////////////

// data: e.g. JSON.stringify({...})
// key: AES key (256 bit Buffer)
// iv: optional initialization vector
// opts: optional mode
// returns: concatenated and Base64 encoded iv + payload
export const encryptDataWithKey = curry((data, key, iv, options) => {
  options = options || {}
  options.padding = U.Iso10126
  const IV = iv || crypto.randomBytes(U.SALT_BYTES)
  const dataBytes = Buffer.from(data, 'utf8')
  const { encryptedBytes, tag } = U.AES.encrypt(dataBytes, key, IV, options)
  const payload = tag
    ? Buffer.concat([tag, IV, encryptedBytes])
    : Buffer.concat([IV, encryptedBytes])
  return payload.toString('base64')
})

// data: base64 string
// key: AES key (256 bit Buffer)
// returns; decrypted payload (e.g. a JSON string)
export const decryptDataWithKey = curry((data, key) => {
  const dataHex = Buffer.from(data, 'base64')
  const iv = dataHex.slice(0, U.SALT_BYTES)
  const payload = dataHex.slice(U.SALT_BYTES)
  // @ts-ignore
  return decryptBufferWithKey(payload, iv, key)
})

// sha256 :: Buffer -> Buffer
export const sha256 = (data) => crypto.createHash('sha256').update(data).digest()

// generateMnemonic :: Api -> Promise String
export const generateMnemonic = (api) => {
  return createRng(16, api).then((rng) => BIP39.generateMnemonic(null, rng))
}

// stretchPassword :: password -> salt -> iterations -> keylen -> Task Error Buffer
export const stretchPassword = (password, salt, iterations, keyLenBits) => {
  return satisfyAll([
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required'],
    [salt, 'salt_required'],
    [isNil(keyLenBits) || keyLenBits % 8 === 0, 'key_len_multiple_of_8_required']
  ]).chain(() => pbkdf2Task(password, salt, iterations, keyLenBits))
}

// decryptDataWithPassword :: data -> password -> iterations -> options -> Task Error Buffer
export const decryptDataWithPassword = (data, password, iterations, options) => {
  if (!data) return Task.of(data)
  return satisfyAll([
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required']
  ]).chain(() =>
    // @ts-ignore
    TaskTry(toPayloadIV)(data, options).chain(({ iv, payload }) =>
      stretchPassword(password, iv, iterations, U.KEY_BIT_LEN).chain((key) =>
        // @ts-ignore
        TaskTry(decryptBufferWithKey)(payload, iv, key, options)
      )
    )
  )
}

// encryptDataWithPassword :: String -> String -> Integer -> Task, Error String
export const encryptDataWithPassword = (data, password, iterations) => {
  if (!data) return Task.of(data)
  const salt = crypto.randomBytes(U.SALT_BYTES)
  return satisfyAll([
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required']
  ])
    .chain(() => stretchPassword(password, salt, iterations, U.KEY_BIT_LEN))
    .chain((key) =>
      // @ts-ignore
      TaskTry(encryptDataWithKey)(data, key, salt, { mode: U.AES.CBC })
    )
}

// encryptWallet :: String -> String -> Integer -> String -> Task Error String
export const encryptWallet = curry((data, password, iterations, version) =>
  satisfyAll([
    [data, 'data_required'],
    [is(String, password), 'password_required'],
    [is(Number, iterations) && iterations > 0, 'iterations_required'],
    [version, 'version_required']
  ])
    .chain(() => encryptDataWithPassword(data, password, iterations))
    .map((payload) => JSON.stringify({ payload, pbkdf2_iterations: iterations, version }))
)

// encryptSecPass :: String -> Integer -> String -> String -> Task, Error String
export const encryptSecPass = curry((sharedKey, pbkdf2Iterations, password, message) =>
  encryptDataWithPassword(message, sharedKey + password, pbkdf2Iterations)
)

const checkFailure = curry((pass, fail, str) =>
  str === '' ? fail(new Error('DECRYPT_FAILURE')) : pass(str)
)

export const decryptSecPass = curry((sharedKey, pbkdf2Iterations, password, message) =>
  // @ts-ignore
  decryptDataWithPassword(message, sharedKey + password, pbkdf2Iterations).chain(
    checkFailure(Task.of, Task.rejected)
  )
)

// decryptWalletV1 :: String -> String -> Task Error Object
export const decryptWalletV1 = (password, data) => {
  const decrypt = (i, o) =>
    decryptDataWithPassword(data, password, i, o).chain((p) =>
      safeParse(p, 'v1: wrong_wallet_password')
    )
  // v1: CBC, ISO10126, 10 iterations
  return (
    // @ts-ignore
    decrypt(10)
      // v1: OFB, nopad, 1 iteration
      .orElse(decrypt.bind(null, 1, { mode: U.AES.OFB, padding: U.NoPadding }))
      // v1: OFB, ISO7816, 1 iteration
      // ISO/IEC 9797-1 Padding method 2 is the same as ISO/IEC 7816-4:2005
      .orElse(decrypt.bind(null, 1, { mode: U.AES.OFB, padding: U.Iso97971 }))
      // v1: CBC, ISO10126, 1 iteration
      .orElse(decrypt.bind(null, 1, { mode: U.AES.CBC, padding: U.Iso10126 }))
  )
}

const validateWrapper = (
  wrapper // wrapper
) =>
  satisfyAll([
    [has('payload', wrapper), 'v2v3 wrapper: payload_required'],
    [has('version', wrapper), 'v2v3 wrapper: version_required'],
    [has('pbkdf2_iterations', wrapper), 'v2v3 wrapper: pbkdf2_iterations_required'],
    [
      // @ts-ignore
      propSatisfies((v) => v <= SUPPORTED_ENCRYPTION_VERSION, 'version', wrapper),
      `Wallet version ${wrapper.version} not supported.`
    ]
  ]).map(() => wrapper)

// decryptWalletV2V3 :: String -> String -> Task Error Object
export const decryptWalletV2V3 = (password, data) => {
  return (
    safeParse(data, 'v2v3: wrong_wrapper')
      .chain(validateWrapper)
      // v2/v3: CBC, ISO10126, iterations in wrapper
      .chain((w) =>
        decryptDataWithPassword(w.payload, password, w.pbkdf2_iterations, {
          mode: U.AES.CBC
        })
      )
      .chain((p) => safeParse(p, 'v2v3: wrong_wallet_password'))
  )
}

export const decryptWallet = curry((password, data) =>
  decryptWalletV1(password, data).orElse((result) => {
    return result === 'v1: wrong_wallet_password'
      ? Task.rejected(result)
      : decryptWalletV2V3(password, data)
  })
)

// @ts-ignore
window.decryptWallet = decryptWallet

// @ts-ignore
window.decryptWalletV2V3 = decryptWalletV2V3

// @ts-ignore
window.decryptDataWithPassword = decryptDataWithPassword

export const derivePubFromPriv = (priv) => {
  return Bitcoin.ECPair.fromPrivateKey(priv).publicKey
}

export const deriveSharedSecret = (priv, pub) => {
  const c = curve.getCurveByName('secp256k1')
  const privNumber = BigInteger.fromBuffer(priv)

  const p = curve.Point.decodeFrom(c, pub)
  const m = p.multiply(privNumber)

  const encoded = m.getEncoded(true)
  const hashed = sha256(encoded)

  return hkdf(hashed, 32, { hash: 'SHA-256' })
}

export const encryptAESGCM = (key, msg) => {
  const IV = crypto.randomBytes(12)
  const options = { mode: U.AES.GCM }
  const encryptedBytes = U.AES.encrypt(msg, key, IV, options)
  return Buffer.concat([IV, encryptedBytes])
}

export const decryptAESGCM = (key, msg) => {
  const IV = msg.slice(0, 12)
  const dataBytes = msg.slice(12)
  const options = { mode: U.AES.GCM }
  return U.AES.decrypt(dataBytes, key, IV, options)
}
