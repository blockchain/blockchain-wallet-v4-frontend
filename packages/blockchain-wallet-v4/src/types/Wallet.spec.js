import * as R from 'ramda'
import { Address, Wallet, AddressMap, serializer } from './index'
import * as crypto from '../walletCrypto/index'

const walletFixture = require('./__mocks__/wallet.v3')
const walletFixtureSecpass = require('./__mocks__/wallet.v3-secpass')
const secpass = 'secret'

describe('Wallet', () => {
  const wallet = Wallet.fromJS(walletFixture)
  const walletSecpass = Wallet.fromJS(walletFixtureSecpass)

  crypto.encryptDataWithKey = (data, key, iv) => (data ? `enc<${data}>` : null)

  describe('toJS', () => {
    it('should return the correct object', () => {
      let result = Wallet.toJS(wallet)
      expect(result).toEqual(walletFixture)
    })
  })

  describe('selectors', () => {
    it('should throw if it does not receive a Wallet instance', () => {
      expect(() => Wallet.select(Wallet.guid, {})).toThrow(TypeError)
    })

    it('should select guid', () => {
      expect(wallet.guid).toEqual(walletFixture.guid)
      expect(Wallet.selectGuid(wallet)).toEqual(walletFixture.guid)
    })

    it('should select iterations', () => {
      expect(Wallet.selectIterations(wallet)).toEqual(
        walletFixture.options.pbkdf2_iterations
      )
    })

    it('should select addresses', () => {
      let addressesJS = R.compose(
        AddressMap.toJS,
        Wallet.selectAddresses
      )(wallet)
      expect(addressesJS).toEqual(walletFixture.keys)
    })

    it('should select if is double encrypted', () => {
      expect(Wallet.isDoubleEncrypted(wallet)).toEqual(false)
      expect(Wallet.isDoubleEncrypted(walletSecpass)).toEqual(true)
    })
  })

  describe('importLegacyAddress', () => {
    const n = walletFixture.keys.length
    const address = Address.fromJS({
      addr: '1CC3X2gu58d6wXUWMffpuzN9JAfTUWu4Kj',
      priv: '5Kb8kLf9zgWQnogidDA76MzPL6TsZZY36hWXMssSzNydYXYB9KF'
    })

    it('should add an unencrypted address', () => {
      Wallet.importLegacyAddress(
        wallet,
        address.priv,
        0,
        undefined,
        undefined,
        {}
      ).fork(
        failure => expect(failure).toEqual(undefined),
        withNewAddress => {
          let addresses = Wallet.selectAddresses(withNewAddress)
          expect(addresses.size).toEqual(n + 1)
          const newAddr = R.compose(
            AddressMap.selectAddress('1CC3X2gu58d6wXUWMffpuzN9JAfTUWu4Kj'),
            Wallet.selectAddresses
          )(withNewAddress)
          expect(newAddr.addr).toEqual(address.addr)
          expect(newAddr.priv).toEqual(
            'GibbqZhygNfhbfz4fb2vDg19Ym5v696w52iqZEQySHTw'
          )
        }
      )
    })

    it('should add a double encrypted address', () => {
      Wallet.importLegacyAddress(
        wallet,
        address.priv,
        0,
        'secret',
        undefined,
        {}
      ).fork(
        failure => expect(failure).toEqual(undefined),
        withNewAddress => {
          let addresses = Wallet.selectAddresses(withNewAddress)
          expect(addresses.size).toEqual(n + 1)
          const newAddr = R.compose(
            AddressMap.selectAddress('1CC3X2gu58d6wXUWMffpuzN9JAfTUWu4Kj'),
            Wallet.selectAddresses
          )(withNewAddress)
          expect(newAddr.addr).toEqual(address.addr)
          expect(newAddr.priv).toEqual(
            'GibbqZhygNfhbfz4fb2vDg19Ym5v696w52iqZEQySHTw'
          ) // << TODO this is not encrypted?
        }
      )
    })
  })

  describe('setAddressLabel', () => {
    it('should set a new address label', () => {
      let addr = '14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ'
      let withNewLabel = Wallet.setLegacyAddressLabel(addr, 'new_label', wallet)
      expect(wallet.addresses.get(addr).label).toEqual('labeled_imported')
      expect(withNewLabel.addresses.get(addr).label).toEqual('new_label')
    })
  })

  describe('isValidSecondPwd', () => {
    it('should be valid for an unencrypted wallet', () => {
      let isValid = Wallet.isValidSecondPwd(null, wallet)
      expect(isValid).toEqual(true)
    })

    it('should detect a valid second password', () => {
      let isValid = Wallet.isValidSecondPwd(secpass, walletSecpass)
      expect(isValid).toEqual(true)
    })

    it('should detect an invalid second password', () => {
      let secpass = 'wrong_secpass'
      let isValid = Wallet.isValidSecondPwd(secpass, walletSecpass)
      expect(isValid).toEqual(false)
    })
  })

  describe('encrypt', () => {
    it('should encrypt', done => {
      Wallet.encrypt('secret', wallet).fork(done, encrypted => {
        let [before, after] = [wallet, encrypted].map(Wallet.selectAddresses)
        let enc = crypto.encryptDataWithKey
        let success = R.zip(before, after).every(
          ([b, a]) => enc(b.priv) === a.priv
        )
        expect(success).toEqual(true)
        done()
      })
    })
  })

  describe('decrypt', () => {
    it('should decrypt', done => {
      Wallet.decrypt('secret', walletSecpass).fork(done, decrypted => {
        expect(Wallet.toJS(decrypted)).toEqual(walletFixture)
        done()
      })
    })

    it('should fail when given an incorrect password', done => {
      Wallet.decrypt('wrong', walletSecpass).fork(error => {
        expect(R.is(Error, error)).toEqual(true)
        expect(error.message).toEqual('INVALID_SECOND_PASSWORD')
        done()
      }, done)
    })
  })

  describe('serializer', () => {
    it('compose(reviver, replacer) should be identity', () => {
      const string = JSON.stringify(wallet)
      const newWallet = JSON.parse(string, serializer.reviver)
      expect(newWallet).toEqual(wallet)
    })
    it('compose(replacer, reviver) should be identity', () => {
      const string = JSON.stringify(wallet)
      const newWallet = JSON.parse(string, serializer.reviver)
      const string2 = JSON.stringify(newWallet)
      expect(string2).toEqual(string)
    })
  })

  // describe('createNew', () => {
  //   const { mnemonic } = walletNewFixture

  //   it('should create a new wallet', () => {
  //     let { guid, sharedKey } = walletNewFixture.wallet
  //     let wallet = Wallet.createNew(guid, sharedKey, mnemonic)
  //     expect(Wallet.toJS(wallet)).toEqual(walletNewFixture.wallet)
  //   })
  // })
})
