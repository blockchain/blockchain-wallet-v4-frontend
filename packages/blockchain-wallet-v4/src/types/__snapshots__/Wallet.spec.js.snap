// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet newHDAccount 1`] = `
Object {
  "address_book": Object {},
  "addresses": Object {
    "14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ": Object {
      "addr": "14mQxLtEagsS8gYsdWJbzthFFuPDqDgtxQ",
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
      "created_time": 1492721432222,
      "label": "labeled_imported",
      "priv": "BpD2ZuJjZ8PJPpDX6ZmsKFsXHkL7XV3dt385zghMfF6C",
      "tag": 0,
    },
    "19XmKRY66VnUn5irHAafyoTfiwFuGLUxKF": Object {
      "addr": "19XmKRY66VnUn5irHAafyoTfiwFuGLUxKF",
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
      "created_time": 1492721419269,
      "label": "",
      "priv": "38W3vsxt246Lhawvz81y2HQVnsJrBUX87jSTUUnixPsE",
      "tag": 0,
    },
    "1JD73aGSdeqKUjJ4ntP4eCyUiuZ3ogJE1u": Object {
      "addr": "1JD73aGSdeqKUjJ4ntP4eCyUiuZ3ogJE1u",
      "created_device_name": "javascript_web",
      "created_device_version": "3.0",
      "created_time": 1492721461228,
      "label": "",
      "priv": null,
      "tag": 0,
    },
  },
  "double_encryption": false,
  "guid": "50dae286-e42e-4d67-8419-d5dcc563746c",
  "hd_wallets": Array [
    Object {
      "accounts": Array [
        Object {
          "address_labels": Object {
            "0": Object {
              "index": 0,
              "label": "labeled_address",
            },
          },
          "archived": false,
          "cache": Object {
            "changeAccount": "xpub6F41z8MqNcJMwmeUExdCv7UXvYBEgQB29SWq9jyxuZ7WefmSTWcwXB6NRAJkGCkB3L1Eu4ttzWnPVKZ6REissrQ4i6p8gTi9j5YwDLxmZ8p",
            "receiveAccount": "xpub6F41z8MqNcJMvKQgAd5QE2QYo32cocYigWp1D8726ykMmaMqvtqLkvuL1NqGuUJvU3aWyJaV2J4V6sD7Pv59J3tYGZdYRSx8gU7EG8ZuPSY",
          },
          "index": 0,
          "label": "My Bitcoin Wallet",
          "xpriv": "xprv9yL1ousLjQQzGNBAYykaT8J3U626NV6zbLYkRv8rvUDpY4f1RnrvAXQneGXC9UNuNvGXX4j6oHBK5KiV2hKevRxY5ntis212oxjEL11ysuG",
          "xpub": "xpub6CKNDRQEZmyHUrFdf1HapGEn27ramwpqxZUMEJYUUokoQrz9yLBAiKjGVWDuiCT39udj1r3whqQN89Tar5KrojH8oqSy7ytzJKW8gwmhwD3",
        },
        Object {
          "address_labels": Object {},
          "archived": false,
          "cache": Object {
            "changeAccount": "xpub68Gmy5EVb2BdHTYHpekwGdcbBWax19w9HwA2DaADYvuCSSgt4YAErxxSN1KWSnmyqkwRNbnTj3XiUBKmHeC8rTjLRPjSULcDKQQgfgJDppq",
            "receiveAccount": "xpub68Gmy5EVb2BdFbj2LpWrk1M7obNuaPTpT5oh9QCCo5sRfqSHVYWex97WpDZzszdzHzxXDAzPLVSwybe4uPYkSk4G3gnrPqqkV9RyNzAcNJ1",
          },
          "index": 1,
          "label": "label",
          "xpriv": "xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi",
          "xpub": "xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8",
        },
      ],
      "default_account_idx": 0,
      "mnemonic_verified": false,
      "passphrase": "",
      "seedHex": "6a4d9524d413fdf69ca1b5664d1d6db0",
    },
  ],
  "metadataHDNode": "xprv9tygGQP8be7uzNm5Czuy41juTK9pUKnWyZtDxgbmSEcCYa9VdvvtSknEyiKitqqm2TMv14NjXPQ68XLwSdH6Scc5GwXoZ31yRZZysxhVGU7",
  "options": Object {
    "fee_per_kb": 10000,
    "html5_notifications": false,
    "logout_time": 600000,
    "pbkdf2_iterations": 5000,
  },
  "sharedKey": "8a260b2b-5257-4357-ac56-7a7efca323ea",
  "tx_names": Array [],
  "tx_notes": Object {},
}
`;
